lib/datadog/statsd/serialization/tag_serializer.rb

```diff
@@ -13,18 +13,21 @@ def initialize(global_tags = [], env = ENV)
 
           # Convert to tag list and set
           @global_tags = to_tags_list(global_tags)
+          @global_tags_formatted = @global_tags.join(',') if @global_tags.any?
         end
 
         def format(message_tags)
-          # fast return global tags if there's no message_tags
-          # to avoid more allocations
-          tag_list =  if message_tags && message_tags.any?
-                        global_tags + to_tags_list(message_tags)
-                      else
-                        global_tags
-                      end
+          if !message_tags || message_tags.empty?
+            return @global_tags_formatted
+          end
+
+          tags = if @global_tags_formatted
+                   [@global_tags_formatted, to_tags_list(message_tags)]
```

Comment:
I was surprised to find out that `Array#join` flattens the provided array:
```ruby
[1,[2,3]].join
# => "123"
```
This actually came in handy, as it has proven faster than adding `@global_tags_formatted` to the array returned by `to_tags_list(message_tags)` or concatenating the result of `to_tags_list(message_tags)` to an existing array.

Reply:
That is interesting; it's like a `flatten` then regular `join` operation.